packages/sol-compiler/src/compiler.ts

```diff
@@ -107,97 +150,100 @@ export class Compiler {
+            const input = versionToInputs[solcVersion];
+            logUtils.log(
+                `Compiling ${input.contractsToCompile.length} contracts (${
+                    input.contractsToCompile
+                }) with Solidity v${solcVersion}...`,
+            );
 
-        if (!_.isUndefined(compiled.errors)) {
-            const SOLIDITY_WARNING = 'warning';
-            const errors = _.filter(compiled.errors, entry => entry.severity !== SOLIDITY_WARNING);
-            const warnings = _.filter(compiled.errors, entry => entry.severity === SOLIDITY_WARNING);
-            if (!_.isEmpty(errors)) {
-                errors.forEach(error => {
-                    const normalizedErrMsg = getNormalizedErrMsg(error.formattedMessage || error.message);
-                    logUtils.log(chalk.red(normalizedErrMsg));
-                });
-                process.exit(1);
-            } else {
-                warnings.forEach(warning => {
-                    const normalizedWarningMsg = getNormalizedErrMsg(warning.formattedMessage || warning.message);
-                    logUtils.log(chalk.yellow(normalizedWarningMsg));
-                });
+            const { solcInstance, fullSolcVersion } = await Compiler._getSolcAsync(solcVersion);
+
+            const compilerOutput = this._compile(solcInstance, input.standardInput);
+
+            for (const contractPath of input.contractsToCompile) {
+                await this._verifyAndPersistCompiledContractAsync(
+                    contractPath,
+                    contractPathToData[contractPath],
+                    fullSolcVersion,
+                    compilerOutput,
+                );
             }
         }
-        const compiledData = compiled.contracts[contractSource.path][contractName];
+    }
+    private _shouldCompile(contractData: ContractData): boolean {
+        if (_.isUndefined(contractData.currentArtifactIfExists)) {
+            return true;
+        } else {
+            const currentArtifact = contractData.currentArtifactIfExists as ContractArtifact;
+            const isUserOnLatestVersion = currentArtifact.schemaVersion === constants.LATEST_ARTIFACT_VERSION;
+            const didCompilerSettingsChange = !_.isEqual(currentArtifact.compiler.settings, this._compilerSettings);
+            const didSourceChange = currentArtifact.sourceTreeHashHex !== contractData.sourceTreeHashHex;
+            return !isUserOnLatestVersion || didCompilerSettingsChange || didSourceChange;
+        }
+    }
+    private async _verifyAndPersistCompiledContractAsync(
+        contractPath: string,
+        contractMetadata: ContractData,
+        fullSolcVersion: string,
+        compilerOutput: solc.StandardOutput,
+    ): Promise<void> {
+        const contractName = contractMetadata.contractName;
+        const sourceTreeHashHex = contractMetadata.sourceTreeHashHex;
+        const currentArtifactIfExists = contractMetadata.currentArtifactIfExists;
```

Comment:
I'm not a huge fan of this, it feels deceptive of what all that this method takes in as parameters. Instead let's make this methods signature:

```
    private async _verifyAndPersistCompiledContractAsync(
        contractPath: string,
        contractName: string,
        sourceTreeHashHex: string,
        artifact: TYPE,
        fullSolcVersion: string,
        compilerOutput: solc.StandardOutput,
    ): Promise<void> {
```

Reply:
I agree, it is deceptive, especially since it's just a one-off bundle of data that doesn't get re-used anywhere else.